Update websockets

bishoy-at-pieces opened the pull request:
closes #4

bishoy-at-pieces committed:
add websockets

bishoy-at-pieces committed:
update annotations and snapshots

bishoy-at-pieces committed:
add docs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/mason-at-pieces/pieces-copilot-python-sdk?pullRequest=5) <br/>All committers have signed the CLA.

bishoy-at-pieces committed:
fix typo

bishoy-at-pieces committed:
fix caching.md

bishoy-at-pieces committed:
fix @classmethod issue and import error

bishoy-at-pieces committed:
add validations and fix annotations

bishoy-at-pieces committed:
fix import error in annotations

bishoy-at-pieces committed:
added connector api

bishoy-at-pieces committed:
fix import error

bishoy-at-pieces committed:
fix docs

bishoy-at-pieces committed:
fix health websocket

mason-at-pieces committed:
Merge branch 'main' into update-websockets

mason-at-pieces commented:
@Bishoy-at-pieces could you please add in a file that displays the use cases for the added methods?

I have written tests for the original methods and will merge them in but just need to ensure that this all works. Also need to ensure that we are good with the experience that this gives a developer

bishoy-at-pieces commented:
docs/caching.md
docs/websockets.md
@mason-at-pieces

bishoy-at-pieces commented:
Bow it is much much easier to interact with assets and conversations is cached so anything related to conversation instead of fetching the api you should use the cache

bishoy-at-pieces committed:
added the create asset and the docs

bishoy-at-pieces committed:
fix docs

bishoy-at-pieces committed:
add docs comments to the websockets

bishoy-at-pieces committed:
fix application

bishoy-at-pieces committed:
remove the get_conversations

bishoy-at-pieces committed:
fix docs

bishoy-at-pieces committed:
add asset wrapper

sambhavnoobcoder committed:
initialise tests

initialise the tests directory for the sdk

bishoy-at-pieces committed:
update add copilot ask

bishoy-at-pieces committed:
refactor BasicAsset class

bishoy-at-pieces committed:
add the available_models_names property

bishoy-at-pieces committed:
fix asset setter

bishoy-at-pieces committed:
add property to the to_image method

bishoy-at-pieces committed:
remove import error

bishoy-at-pieces committed:
fix annotations and create asset

bishoy-at-pieces committed:
Refactor assets

bishoy-at-pieces committed:
add chat class

bishoy-at-pieces committed:
add the message model

bishoy-at-pieces committed:
add doc string to the copilot

bishoy-at-pieces committed:
fix the message and chat

bishoy-at-pieces committed:
add doc comment to the chat model

bishoy-at-pieces committed:
add docs

bishoy-at-pieces committed:
fix url

bishoy-at-pieces committed:
fix the brackets

bishoy-at-pieces committed:
remove extra brackets

bishoy-at-pieces committed:
remove the chat_id extra

bishoy-at-pieces committed:
add reclassify asset and refactor asset

bishoy-at-pieces committed:
fix the classification if it is an image

bishoy-at-pieces committed:
fix the raw method

bishoy-at-pieces committed:
fix docs

bishoy-at-pieces committed:
fix import error

bishoy-at-pieces committed:
fix attribute error

bishoy-at-pieces committed:
add on exit to close all websockets

bishoy-at-pieces committed:
fix creating a lot of threads issue

bishoy-at-pieces committed:
add dunder methods

bishoy-at-pieces committed:
fix the basic message model

bishoy-at-pieces committed:
fix the copilot docs

bishoy-at-pieces committed:
add the question endpoint

bishoy-at-pieces committed:
add prompt pipeline

bishoy-at-pieces committed:
refactor streamed identifiers

bishoy-at-pieces committed:
fix copilot method and made it sync

bishoy-at-pieces committed:
fix docs for copilot

bishoy-at-pieces committed:
update docs

sambhavnoobcoder committed:
Merge branch 'Bishoy-at-pieces:update-websockets' into update-websockets

sambhavnoobcoder committed:
Create BasicAsset-Testing

added the basic imports

sambhavnoobcoder committed:
debugging modification 

slight modification to the initialisation of the BasicAsset class to allow for some debugging

sambhavnoobcoder committed:
Add pytest fixture for PiecesClient

This commit introduces a pytest fixture that creates and returns a PiecesClient instance. The fixture:

- Has function scope, ensuring a fresh client for each test
- Configures the client with a base URL of 'http://localhost:1000/'
- Allows easy injection of the client into test functions

sambhavnoobcoder committed:
Add setup_asset_snapshot fixture for test isolation

This commit introduces a pytest fixture that manages the AssetSnapshot.identifiers_snapshot:

- Automatically runs for each test (autouse=True)
- Clears identifiers_snapshot before each test
- Prints the snapshot state before and after each test
- Yields control during test execution
- Clears the snapshot again after each test

sambhavnoobcoder committed:
Add mock_asset pytest fixture

This commit introduces a pytest fixture that creates a mock Asset object:

- Uses MagicMock to create a flexible mock object
- Sets up basic properties like id, name, and content
- Configures classification as Python code
- Initializes an empty formats list

sambhavnoobcoder committed:
Add mock_asset_snapshot pytest fixture

This commit introduces a new pytest fixture that sets up a mock AssetSnapshot:

- Utilizes the existing mock_asset fixture
- Adds the mock asset to AssetSnapshot.identifiers_snapshot
- Uses the mock asset's ID as the key in the snapshot
- Prints a confirmation message for debugging
- Returns the AssetSnapshot class

sambhavnoobcoder committed:
Add test for BasicAsset initialization

This commit introduces a new test function that verifies the basic initialization of a BasicAsset:

- Uses mock_asset_snapshot and mock_asset fixtures
- Prints debug information about the AssetSnapshot state and mock asset ID
- Initializes a BasicAsset with the mock asset's ID
- Asserts that the BasicAsset's asset attribute matches the mock asset

sambhavnoobcoder committed:
Add test for BasicAsset raw property

This commit introduces a new test function that verifies the 'raw' property of BasicAsset:

- Utilizes mock_asset_snapshot and mock_asset fixtures
- Initializes a BasicAsset with the mock asset's ID
- Asserts that the 'raw' property returns the expected content ("Test content")

sambhavnoobcoder committed:
Add unit test for BasicAsset.is_image property

- Add unit test to verify BasicAsset.is_image property behavior
- Test covers two scenarios:
  - Asset not classified as an image
  - Asset classified as an image
- Utilize mocking to simulate different asset classifications

sambhavnoobcoder committed:
Add unit test for BasicAsset.classification property

- Create unit test for BasicAsset.classification property
- Verify that the classification property returns the expected value
- Use mock objects to simulate asset data
- Check if the classification matches ClassificationSpecificEnum.PY

sambhavnoobcoder committed:
Add unit test for BasicAsset.raw setter method

- Implement test for editing content using BasicAsset.raw setter
- Mock pieces_client and Format objects to simulate API interactions
- Verify that:
  - format_update_value is called when setting new content
  - The asset's content is updated correctly
- Use patch to mock external dependencies
- Ensure proper API method calls and content updates

sambhavnoobcoder committed:
Add unit test for BasicAsset.name setter method

- Create test for editing asset name using BasicAsset.name setter
- Mock pieces_client and asset_api to simulate API interactions
- Verify that:
  - The asset's name is updated correctly
  - asset_update method is called when setting a new name
- Use patch to mock external dependencies
- Ensure proper API method calls and name updates

sambhavnoobcoder committed:
Add unit test for BasicAsset.name property

- Implement test for BasicAsset.name property getter
- Verify two scenarios:
  - When asset has a name set
  - When asset name is None (fallback to default name)
- Use mock objects to simulate asset data
- Ensure correct name retrieval and default name fallback

sambhavnoobcoder committed:
Add unit test for BasicAsset.description property

- Create test for BasicAsset.description property getter
- Mock asset annotations to simulate different annotation types
- Verify that:
  - The correct description is returned from annotations
  - Only annotations of type "DESCRIPTION" are considered
- Use MagicMock to create mock Annotation objects
- Ensure proper filtering and retrieval of description from annotations

sambhavnoobcoder committed:
Add unit test for BasicAsset.annotations property

- Implement test for BasicAsset.annotations property getter
- Mock asset annotations using MagicMock
- Verify that:
  - The annotations property returns the correct list of annotations
  - The returned annotations match the mocked annotations
- Ensure proper retrieval of annotations from the asset

sambhavnoobcoder committed:
Add unit test for BasicAsset.delete method

- Create test for BasicAsset.delete method
- Mock pieces_client and assets_api to simulate API interactions
- Verify that:
  - The delete method is called successfully
  - assets_delete_asset API method is called with the correct asset ID
- Use patch to mock external dependencies
- Ensure proper API method call for asset deletion

sambhavnoobcoder committed:
Add unit test for BasicAsset.create class method

- Implement test for BasicAsset.create static method
- Mock pieces_client and assets_api to simulate asset creation
- Verify that:
  - The create method returns the correct new asset ID
  - assets_create_new_asset API method is called
- Use patch to mock external dependencies
- Ensure proper API method call and return value for asset creation

sambhavnoobcoder committed:
Delete tests/README.md

 not req hence removed

bishoy-at-pieces committed:
Merge pull request #1 from sambhavnoobcoder/update-websockets

Add unit tests for BasicAsset class

bishoy-at-pieces committed:
update the test name

bishoy-at-pieces committed:
fix decoding error

bishoy-at-pieces committed:
implantation of singleton class

bishoy-at-pieces committed:
remove singleton

bishoy-at-pieces committed:
fix quotes

bishoy-at-pieces committed:
add the ensure_initialization

bishoy-at-pieces committed:
update docs

sambhavnoobcoder committed:
Create test_basic_chat.py

- importing basic dependencies

sambhavnoobcoder committed:
file imports

added imports from the files

bishoy-at-pieces committed:
fix some naming stuff

bishoy-at-pieces committed:
update docs

sambhavnoobcoder committed:
imports from basic_identifier

added imports to the following classes : 
- Basic
- BasicChat
- BasicMessage

sambhavnoobcoder committed:
Added imports from the streamed_identifiers

this commit adds the imports of the classes : 
- StreamedIdentifiersCache
- ConversationsSnapshot

bishoy-at-pieces committed:
Update basic_message.md

sambhavnoobcoder committed:
Add setup fixture for TestBasicChat class

This commit introduces a setup fixture for the TestBasicChat class:

- Create a mock conversation with a test ID and empty messages
- Set the conversation name to "Test Conversation"
- Add the mock conversation to ConversationsSnapshot.identifiers_snapshot
- Mock the ConversationsSnapshot.pieces_client

sambhavnoobcoder committed:
Add test for BasicChat initialization with valid ID

This commit introduces a new test method test_init_valid_id in the TestBasicChat class:

- Creates a BasicChat instance with a valid ID "test_id"
- Asserts that the chat's ID matches the provided ID
- Asserts that the chat's name is correctly set to "Test Conversation"

sambhavnoobcoder committed:
Add test for BasicChat initialization with invalid ID

This commit adds a new test method test_init_invalid_id to the TestBasicChat class:

- Attempts to create a BasicChat instance with an invalid ID "invalid_id"
- Uses pytest.raises to assert that a ValueError is raised
- Checks that the error message contains "Conversation not found"

sambhavnoobcoder committed:
Add test for BasicChat name property

This commit introduces a new test method test_name_property in the TestBasicChat class:

- Creates a BasicChat instance with ID "test_id"
- Verifies the initial name is "Test Conversation"
- Changes the name to "New Name"
- Asserts that the name property is updated correctly
- Checks that the conversation_update method of the API client is called once

sambhavnoobcoder committed:
Add test for BasicChat name property with default value

This commit adds a new test method test_name_property_default to the TestBasicChat class:

- Sets the mock conversation's name to None
- Creates a BasicChat instance with ID "test_id"
- Asserts that the chat's name defaults to "New Conversation"

sambhavnoobcoder committed:
Add comprehensive test for BasicChat messages method

This commit introduces a new test method test_messages in the TestBasicChat class:

- Uses @patch.object to mock BasicMessage.__init__ and BasicChat._get_message
- Sets up mock conversation with three message indices, including a deleted message
- Creates a BasicChat instance and calls its messages() method
- Asserts that the returned messages list has the correct length
- Verifies that all returned items are instances of BasicMessage
- Checks that BasicMessage.__init__ and _get_message were called the correct number of times
- Ensures _get_message was called with the correct message IDs
- Verifies that BasicMessage.__init__ was called with the results of _get_message

sambhavnoobcoder committed:
Add test for BasicChat annotations property

This commit adds a new test method test_annotations_property to the TestBasicChat class:

- Creates a mock annotations object with an iterable containing two annotations
- Sets the mock annotations on the test conversation in ConversationsSnapshot
- Instantiates a BasicChat object with the test ID
- Asserts that the chat's annotations property returns the expected list of annotations

sambhavnoobcoder committed:
Add test for BasicChat annotations property when annotations are None

This commit introduces a new test method test_annotations_property_none to the TestBasicChat class:

- Sets the annotations of the test conversation in ConversationsSnapshot to None
- Creates a BasicChat instance with the test ID
- Asserts that the chat's annotations property returns None

sambhavnoobcoder committed:
Add test for BasicChat delete method

This commit adds a new test method test_delete to the TestBasicChat class:

- Creates a BasicChat instance with the test ID
- Calls the delete method on the chat object
- Asserts that the conversations_delete_specific_conversation method of the API client is called once with the correct conversation ID

sambhavnoobcoder committed:
Add test for BasicChat _edit_conversation static method

This commit introduces a new test method test_edit_conversation to the TestBasicChat class:

- Uses @patch.object to mock ConversationsSnapshot.pieces_client
- Creates a mock conversation object
- Calls the static method BasicChat._edit_conversation with the mock conversation
- Asserts that the conversation_update method of the API client is called once with the correct arguments

sambhavnoobcoder committed:
Final Refactor 

Removed some redundant imports

mason-at-pieces committed:
docs: Fix asset client reference in example

bishoy-at-pieces committed:
fix conversation hanging issue

bishoy-at-pieces committed:
Merge branch 'update-websockets' of https://github.com/Bishoy-at-pieces/pieces-copilot-python-sdk into update-websockets

bishoy-at-pieces committed:
update copilot docs

masnwilliams commented on tests/test_basic_asset.py:
```diff
@@ -0,0 +1,142 @@
+    asset = BasicAsset(mock_asset.id)
+    assert asset.asset == mock_asset
+
+def test_raw_property(mock_asset_snapshot, mock_asset):
```
should be named `test_raw_content_property` since we updated `raw` to `raw_content`

masnwilliams commented on tests/test_basic_asset.py:
```diff
@@ -0,0 +1,142 @@
+
+def test_raw_property(mock_asset_snapshot, mock_asset):
+    asset = BasicAsset(mock_asset.id)
+    assert asset.raw == "Test content"
```
`.raw` need to be updated to `.raw_content`

masnwilliams commented on tests/test_basic_asset.py:
```diff
@@ -0,0 +1,142 @@
+        mock_format.classification.generic = ClassificationGenericEnum.CODE
+        mock_format.fragment = MagicMock()
+        mock_format.fragment.string = MagicMock()
+        mock_format.fragment.string.raw = "Test content"
```
`.raw` need to be updated to `.raw_content`

masnwilliams commented on tests/test_basic_asset.py:
```diff
@@ -0,0 +1,142 @@
+        mock_format.fragment.string.raw = "Test content"
+        mock_format_api.format_snapshot.return_value = mock_format
+        
+        asset.raw = new_content
```
`.raw` need to be updated to `.raw_content`

masnwilliams commented on tests/test_basic_asset.py:
```diff
@@ -0,0 +1,142 @@
+        asset.raw = new_content
+        
+        mock_format_api.format_update_value.assert_called_once()
+        assert mock_format.fragment.string.raw == new_content
```
`.raw` need to be updated to `.raw_content`

masnwilliams commented on tests/test_basic_asset.py:
```diff
@@ -0,0 +1,142 @@
+    mock.id = "test_asset_id"
+    mock.name = "Test Asset"
+    mock.original = MagicMock()
+    mock.original.reference.fragment.string.raw = "Test content"
```
`.raw` need to be updated to `.raw_content`

bishoy-at-pieces replied:
This is raw as it is (not a method it is an asset mock)

masnwilliams commented on tests/test_basic_asset.py:
```diff
@@ -0,0 +1,142 @@
+
+@pytest.fixture(scope="function")
+def pieces_client():
+    return PiecesClient(config={'baseUrl': 'http://localhost:1000/'})
```
would rather us keep urls to not include ending `/`. so this should be just `http://localhost:1000`

shivaypiece commented on docs/basic_message.md:
```diff
@@ -0,0 +1,66 @@
+# BasicMessage
+
```
Describe here how a message is related to a Copilot chat/conversation

shivaypiece commented on docs/copilot.md:
```diff
@@ -0,0 +1,102 @@
+**Args:**
+    query (str): The question to ask.
+    relevant_qgpt_seeds (RelevantQGPTSeeds): Context to the model .
+    pipeline (Optional[QGPTPromptPipeline]): the pipeline to use.
```
Should we document the different pipelines?

bishoy-at-pieces replied:
The pipelines should be already documented on the website docs we can an anchor

bishoy-at-pieces committed:
Merge pull request #2 from sambhavnoobcoder/Basic-Chat-Testing

Adding test for Basic Chat Testing

bishoy-at-pieces committed:
fix test

bishoy-at-pieces committed:
fix docs

bishoy-at-pieces committed:
add the context manager for the copilot

bishoy-at-pieces committed:
fix imports

bishoy-at-pieces committed:
fix the create method

sambhavnoobcoder committed:
Create test_basic_message.py

- basic library imports

sambhavnoobcoder committed:
Add pieces_copilot_sdk imports 

this commit adds imports for :
- Copilot class
- PiecesClient class

sambhavnoobcoder committed:
added imports from the basic_identifier directory 

this commit adds imports for the classes :
- BasicAsset
- Basic
- BasicMessage 
- BasicChat

sambhavnoobcoder committed:
Added imports from streamed_identifiers

This commit adds the classes :
- AssetSnapshot
- StreamedIdentifiersCache
- ConversationsSnapshot

sambhavnoobcoder committed:
Add setup fixture for TestBasicMessage class

- Create mock Pieces client and message objects
- Set mock message properties (id, content, role)
- Configure mock API call for message snapshot

sambhavnoobcoder committed:
Add test_init_valid_id method to TestBasicMessage

- Create BasicMessage instance with valid ID
- Assert message ID is set correctly
- Verify API call is made with correct parameters

sambhavnoobcoder committed:
Add test_init_invalid_id method to TestBasicMessage

- Mock API call to raise an exception
- Test that ValueError is raised with correct message
- Ensure BasicMessage init fails for invalid ID

sambhavnoobcoder committed:
Add test_raw_property method to TestBasicMessage

- Initialize BasicMessage with valid ID
- Verify 'raw' property returns correct message content

sambhavnoobcoder committed:
Add test_raw_setter method to TestBasicMessage

- Test setting new content via 'raw' property
- Verify internal message object is updated
- Ensure API update method is called correctly

sambhavnoobcoder committed:
Add test_role_property method to TestBasicMessage

- Initialize BasicMessage with valid ID
- Verify 'role' property returns correct role value

sambhavnoobcoder committed:
Add test_id_property method to TestBasicMessage

- Create BasicMessage instance with specific ID
- Verify 'id' property returns correct message ID

sambhavnoobcoder committed:
Add test_delete_method to TestBasicMessage

- Create BasicMessage instance
- Call delete() method on the message
- Verify API delete method is called with correct ID

sambhavnoobcoder committed:
Add test_annotations_property_none to TestBasicMessage

- Create BasicMessage instance
- Verify 'annotations' property returns None when not set

sambhavnoobcoder committed:
Add test_annotations_property_with_annotations to TestBasicMessage

- Mock message with annotation
- Create BasicMessage instance
- Verify 'annotations' property returns correct Annotations object
- Check annotation details are correctly populated

sambhavnoobcoder committed:
Add test_description_property_no_annotations to TestBasicMessage

- Create BasicMessage instance without annotations
- Verify 'description' property returns None

sambhavnoobcoder committed:
Add test_description_property_with_description to TestBasicMessage

- Mock message with description annotation
- Create BasicMessage instance
- Verify 'description' property returns correct description text
- Ensure annotation API is called correctly

sambhavnoobcoder committed:
Add test_repr method to TestBasicMessage

- Create BasicMessage instance
- Verify __repr__ method returns correct string representation

sambhavnoobcoder committed:
Add test_eq method to TestBasicMessage

- Create multiple BasicMessage instances with same and different IDs
- Test equality and inequality between messages
- Verify correct behavior with non-message objects
- Ensure API calls are made with correct message IDs

sambhavnoobcoder committed:
Add test_str method to TestBasicMessage

- Create BasicMessage instance
- Verify __str__ method returns correct string representation

sambhavnoobcoder committed:
Add test_hash method to TestBasicMessage

- Create BasicMessage instance
- Verify __hash__ method returns hash of message ID

sambhavnoobcoder committed:
Final Refactor

this commit removes redundant imports from the file

bishoy-at-pieces committed:
Merge pull request #3 from sambhavnoobcoder/Basic-Message-Testing

Add Tests for BasicMessage Class

sambhavnoobcoder committed:
Create test_basic_copilot.py

- Add Basic Imports

sambhavnoobcoder committed:
adding more imports

- Added imports for support for tests

sambhavnoobcoder committed:
Add libraries for threading 

- Threading , web sockets support added

sambhavnoobcoder committed:
pieces_copilot_sdk imports

this commit adds classes :
- Copilot
- PiecesClient

sambhavnoobcoder committed:
added basic_identifier imports 

this commit adds the classes : 
- BasicAsset
- Basic
- BasicMessage
- BasicChat

sambhavnoobcoder committed:
added streamed_identifiers imports 

this commit adds the classes : 
- AssetSnapshot
- StreamedIdentifiersCache
- ConversationsSnapshot

sambhavnoobcoder committed:
Implement BasicCopilotTest setUp method

This commit sets up the necessary mocks and dependencies for the BasicCopilotTest class:

- Create a mock `Client` object with the following properties:
  - `tracked_application` is a mock with `id="mock_app_id"`
  - `model_id` is set to `"mock_model_id"`
  - `qgpt_api` is a mock
- Define a `BasicChat` class for testing purposes
  - The `BasicChat` class has an `__init__` method that takes an `id` parameter
  - The `BasicChat` class creates a `Copilot` instance using the `mock_client`
- Patch the `ConversationsSnapshot.identifiers_snapshot` to provide a mock
  conversation snapshot

sambhavnoobcoder committed:
Add tearDown method to BasicCopilotTest

This commit adds the `tearDown` method to the `BasicCopilotTest` class. The
`tearDown` method is responsible for cleaning up the test environment after each
test is executed.

The key changes include:

- Implement the `tearDown` method
- Call `patch.stopall()` in the `tearDown` method
  - This stops all patches created during the test setup

sambhavnoobcoder committed:
Add unit test for Copilot initialization

This commit adds a unit test for the initialization of the `Copilot` class
within the `BasicCopilotTest` class.

The key changes include:

- Implement the `test_init` method
- Verify that the `copilot` attribute is an instance of the `Copilot` class
- Assert that the `pieces_client` attribute of the `copilot` is equal to the
  `mock_client` created in the `setUp` method
- Ensure that the `_on_message_queue` attribute is an instance of the `Queue`
  class
- Verify that the `ask_stream_ws` attribute is an instance of the `AskStreamWS`
  class
- Assert that the `_chat` attribute is `None`

sambhavnoobcoder committed:
Add unit test for Copilot.ask method

This commit adds a unit test for the `ask` method of the `Copilot` class within
the `BasicCopilotTest` class.

The key changes include:

- Implement the `test_ask` method
- Patch the `AskStreamWS` class to create a mock instance
- Create a mock `QGPTStreamEnum.COMPLETED` output and add it to the
  `_on_message_queue`
- Create a mock for the `send_message` method of the `ask_stream_ws` attribute
- Call the `ask` method of the `copilot` instance and assert the expected
  behavior:
  - The result is a list with one element
  - The result element matches the mock output
  - A `BasicChat` instance is created with the expected conversation ID
  - The `send_message` method of the `ask_stream_ws` is called once

sambhavnoobcoder committed:
Add unit test for Copilot.question method

This commit adds a unit test for the `question` method of the `Copilot` class
within the `BasicCopilotTest` class.

The key changes include:

- Implement the `test_question` method
- Create a mock `QGPTOutput` with the following properties:
  - `text` is set to "Test answer"
  - `answers` is an empty list
- Set the `return_value` of the `mock_client.qgpt_api.question` method to the
  mock `QGPTOutput`
- Call the `question` method of the `copilot` instance and assert the expected
  behavior:
  - The result matches the mock `QGPTOutput`
  - The `mock_client.qgpt_api.question` method is called once with the
    `query` argument

sambhavnoobcoder committed:
Add unit test for Copilot.chats method

This commit adds a unit test for the `chats` method of the `Copilot` class within
the `BasicCopilotTest` class.

The key changes include:

- Implement the `test_chats` method
- Patch the `ConversationsSnapshot.identifiers_snapshot` to include two mock
  conversation IDs, "chat1" and "chat2"
- Call the `chats` method of the `copilot` instance and assert the expected
  behavior:
  - The length of the returned list is 2
  - Each element of the list is an instance of the `BasicChat` class

sambhavnoobcoder committed:
Add unit test for Copilot.chat property

This commit adds a unit test for the `chat` property of the `Copilot` class within
the `BasicCopilotTest` class.

The key changes include:

- Implement the `test_chat_property` method
- Assert that the initial value of the `chat` property is `None`
- Create a `BasicChat` instance with the ID "test_conversation_id" and assign it
  to the `chat` property of the `copilot` instance
- Verify that the `chat` property now holds the correct `BasicChat` instance
- Assert that setting the `chat` property to an invalid value (a string) raises
  a `ValueError`

shivaypiece commented on docs/basic_chat.md:
```diff
@@ -0,0 +1,94 @@
+# Retrieve the messages in the conversation
+messages = chat.messages()
+for message in messages:
+    print(f"Message: {message.raw}")
```
raw_content

sambhavnoobcoder committed:
Final refactor 

- Removed redundant imports 
- Removed import repetitions 
- Cleaned code

bishoy-at-pieces committed:
update docs

bishoy-at-pieces committed:
fix docs basic_chat example

bishoy-at-pieces committed:
Merge pull request #4 from sambhavnoobcoder/Basic-Copilot-Testing

Implement Tests for Copilot Class

bishoy-at-pieces commented:
Should I close that PR and open it again on the new repo??
